// sapper/dist/copy_runtime.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var index = require('./index.js');
var path = require('path');
var path__default = _interopDefault(path);
var fs = require('fs');
var create_manifest_data = require('./create_manifest_data.js');
var fs_utils = require('./fs_utils.js');

function inject_resources(build_json_file, asset_dir) {
    var e_1, _a;
    var file_contents = fs.readFileSync(build_json_file, 'utf8');
    var build = JSON.parse(file_contents);
    var deps = {};
    try {
        for (var _b = index.__values(Object.entries(build.dependencies)), _c = _b.next(); !_c.done; _c = _b.next()) {
            var _d = index.__read(_c.value, 2), key = _d[0], value = _d[1];
            deps[key] = value.filter(function (dep) { return dep.endsWith('.css'); });
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_1) throw e_1.error; }
    }
    fs.readdirSync(asset_dir).forEach(function (file) {
        var file_path = path.resolve(asset_dir, file);
        if (!file.endsWith('.js') || fs.statSync(file_path).isDirectory())
            return;
        var source = fs.readFileSync(file_path, 'utf-8');
        var replaced = source.replace(/(\\?["'])__SAPPER_CSS_PLACEHOLDER:([^"']+?)__\1/g, function (m, quotes, route) {
            var replacement = JSON.stringify(process.env.SAPPER_LEGACY_BUILD && deps[route] ?
                deps[route].map(function (_) { return "legacy/" + _; }) :
                deps[route]);
            // If the quotation marks are escaped, then
            // the source code is in a string literal
            // (e.g., source maps) rather than raw
            // JavaScript source. We need to stringify
            // again and then remove the extra quotation
            // marks so that replacement is correct.
            if (quotes[0] === '\\') {
                replacement = JSON.stringify(replacement);
                replacement = replacement.substring(1, replacement.length - 1);
            }
            return replacement;
        });
        fs.writeFileSync(file_path, replaced);
    });
}

function validate_bundler(bundler) {
    if (!bundler) {
        bundler = (fs.existsSync('rollup.config.js') || fs.existsSync('rollup.config.ts') ? 'rollup' :
            fs.existsSync('webpack.config.js') || fs.existsSync('webpack.config.ts') ? 'webpack' : null);
        if (!bundler) {
            // TODO remove in a future version
            deprecate_dir('rollup');
            deprecate_dir('webpack');
            throw new Error("Could not find a configuration file for rollup or webpack");
        }
    }
    if (bundler !== 'rollup' && bundler !== 'webpack') {
        throw new Error("'" + bundler + "' is not a valid option for --bundler \u2014 must be either 'rollup' or 'webpack'");
    }
    return bundler;
}
function deprecate_dir(bundler) {
    try {
        var stats = fs.statSync(bundler);
        if (!stats.isDirectory())
            return;
    }
    catch (err) {
        // do nothing
        return;
    }
    // TODO link to docs, once those docs exist
    throw new Error("As of Sapper 0.21, build configuration should be placed in a single " + bundler + ".config.js file");
}

function copy_shimport(dest) {
    fs.writeFileSync(dest + "/client/shimport@" + create_manifest_data.version + ".js", fs.readFileSync(require.resolve('shimport/index.js')));
}

var runtime = [
    'index.d.ts',
    'app.mjs',
    'server.mjs',
    'internal/shared.mjs',
    'internal/layout.svelte',
    'internal/error.svelte'
].map(function (file) { return ({
    file: file,
    source: fs.readFileSync(path.join(__dirname, "../runtime/" + file), 'utf-8')
}); });
function copy_runtime(output) {
    runtime.forEach(function (_a) {
        var file = _a.file, source = _a.source;
        fs_utils.mkdirp(path.dirname(output + "/" + file));
        fs.writeFileSync(output + "/" + file, source);
    });
}

exports.copy_runtime = copy_runtime;
exports.copy_shimport = copy_shimport;
exports.inject_resources = inject_resources;
exports.validate_bundler = validate_bundler;
